packages/react-native-gesture-handler/src/components/Pressable/StateMachine.tsx

```diff
@@ -0,0 +1,46 @@
+import { PressableEvent } from './PressableProps';
+
+interface StepDefinition {
+  signal: string;
+  callback?: (event: PressableEvent) => void;
+}
+
+class StateMachine {
+  private steps: StepDefinition[];
+  private currentStepIndex: number;
+  private eventPayload: PressableEvent | null;
+
+  constructor(steps: StepDefinition[]) {
+    this.steps = steps;
+    this.currentStepIndex = 0;
+    this.eventPayload = null;
+  }
+
+  public reset() {
+    this.currentStepIndex = 0;
+    this.eventPayload = null;
+  }
+
+  public sendEvent(eventName: string, eventPayload?: PressableEvent) {
+    const step = this.steps[this.currentStepIndex];
+    this.eventPayload = eventPayload || this.eventPayload;
+
+    if (step.signal !== eventName) {
+      if (this.currentStepIndex > 0) {
+        // retry with position at index 0
+        this.reset();
+        this.sendEvent(eventName, eventPayload);
+      }
+      return;
+    }
+
+    this.eventPayload && step.callback?.(this.eventPayload);
+    this.currentStepIndex++;
+
+    if (this.currentStepIndex === this.steps.length) {
+      this.reset();
+    }
```

Comment:
I don't like mixing these 2 conventions in one place. Why don't we change it to 

```suggestion
    if (this.eventPayload && step.callback) {
      step.callback(this.eventPayload);
    }

    this.currentStepIndex++;

    if (this.currentStepIndex === this.steps.length) {
      this.reset();
    }
```

Reply:
Makes sense, changed in [9371db2](https://github.com/software-mansion/react-native-gesture-handler/pull/3571/commits/9371db28f0c47598c5dd0e657d58d984bb2e93ba)